integration/operator_test.go

```diff
@@ -0,0 +1,235 @@
+
+func setUp(cs kubernetes.Interface) error {
+	if err := createGSNamespace(cs); err != nil {
+		return err
+	}
+
+	if err := installVault(cs); err != nil {
+		return err
+	}
+
+	if err := installCertOperator(cs); err != nil {
+		return err
+	}
+	return nil
+}
+
+func tearDown(cs kubernetes.Interface) {
+	runCmd("helm delete vault --purge")
+	runCmd("helm delete cert-operator --purge")
+	runCmd("helm delete cert-resource-lab --purge")
+	cs.CoreV1().Namespaces().Delete("giantswarm", &metav1.DeleteOptions{})
+	cs.ExtensionsV1beta1().ThirdPartyResources().Delete(certificatetpr.Name, &metav1.DeleteOptions{})
+}
+
+func createGSNamespace(cs kubernetes.Interface) error {
+	// check if the namespace already exists
+	_, err := cs.CoreV1().Namespaces().Get("giantswarm", metav1.GetOptions{})
+	if err == nil {
+		return nil
+	}
+
+	namespace := &v1.Namespace{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "giantswarm",
+		},
+	}
+	_, err = cs.CoreV1().Namespaces().Create(namespace)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	return waitFor(activeNamespaceFunc(cs, "giantswarm"))
+}
+
+func installVault(cs kubernetes.Interface) error {
+	if err := runCmd("helm registry install quay.io/giantswarm/vaultlab-chart:stable -- --set vaultToken=${VAULT_TOKEN} -n vault"); err != nil {
+		return microerror.Mask(err)
+	}
+
+	return waitFor(runningPodFunc(cs, "default", "app=vault"))
+}
+
+func installCertOperator(cs kubernetes.Interface) error {
+	certOperatorChartValuesEnv := os.ExpandEnv(certOperatorChartValues)
+	if err := ioutil.WriteFile(certOperatorValuesFile, []byte(certOperatorChartValuesEnv), os.ModePerm); err != nil {
+		return err
+	}
+	if err := runCmd("helm registry install quay.io/giantswarm/cert-operator-chart@1.0.0-${CIRCLE_SHA1} -- -n cert-operator --values " + certOperatorValuesFile); err != nil {
+		return err
+	}
+
+	return waitFor(tprFunc(cs, certificatetpr.Name))
+}
+
+func runCmd(cmdStr string) error {
+	cmdEnv := os.ExpandEnv(cmdStr)
+	fields := strings.Fields(cmdEnv)
+	cmd := exec.Command(fields[0], fields[1:]...)
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stdout
+
+	return cmd.Run()
+}
+
+func waitFor(f func() error) error {
+	timeout := time.After(defaultDeadline * time.Second)
+	ticker := backoff.NewTicker(backoff.NewExponentialBackOff())
+
+	for {
+		select {
+		case <-timeout:
+			ticker.Stop()
+			return fmt.Errorf("wait timeout")
+		case <-ticker.C:
+			if err := f(); err == nil {
+				return nil
+			}
+		}
+	}
+}
+
+func runningPodFunc(cs kubernetes.Interface, namespace, labelSelector string) func() error {
+	return func() error {
+		pods, err := cs.CoreV1().Pods(namespace).List(metav1.ListOptions{
+			LabelSelector: labelSelector,
+		})
+		if err != nil {
+			return microerror.Mask(err)
+		}
+		if len(pods.Items) != 1 {
+			return fmt.Errorf("unexpected number of pods")
+		}
+		pod := pods.Items[0]
+		phase := pod.Status.Phase
+		if phase != v1.PodRunning {
+			return fmt.Errorf("unexpected pod status phase " + string(phase))
```

Comment:
It would be nice to follow our error handling conventions and create error types in `error.go` and use them in the code. There are sometimes quite generic errors like `executionFailedError` which are then annotated via `microerror.Maskf` in case more information should be transported.

Reply:
Sure, i've added a error.go file, ptal.